### DISCUSSION TITLE
Successive setImage() calls

### DISCUSSION THREAD

**Discussion by lukasrosenke** [onevcat/Kingfisher#1646] [Q&A] [answered]
Lets say i call `self.imageView.kf.setImage()` twice within a very short span of time (for the same url). If the download on the first call is through it uses the cached version for the second call (or atleast thats what iam assuming). 
What happens though, if the download/ caching process has not finished and a second call is made ?

**Comment by onevcat** [comment, marked as answer]
> If the download on the first call is through it uses the cached version for the second call (or atleast thats what iam assuming)

Yes, you are right on this.

> What happens though, if the download/ caching process has not finished and a second call is made?

Kingfisher is clever enough to not download the same image multiple times, and it will remember the setting method of last image to use it. 

For your case:

1. The download request of the second call will not start since it is from the same URL and network setting. The retrieving image task will be added to Kingfisher for later use though.
2. When the first download request finishes (either with a success value or failure), the data will be delivered to both "tasks" to convert to an image, to process with their processors (if you are using any), and stored in the cache.
3. Then both of the calls will receive a completion handler. The first handler fails with an error [`notCurrentSourceTask`](https://github.com/onevcat/Kingfisher/blob/master/Sources/General/KingfisherError.swift#L201) and the second handler with a successful result.

```swift
imageView.kf.setImage(with: url) { r in
    switch r {
    case .success: break
    case .failure(let e): e.isNotCurrentTask // true
    }
}

imageView.kf.setImage(with: url) { r in
    print(r) // success
}
```

**Reply by lukasrosenke** [reply]
Thank you very much for this detailed answer.
